langwatch/src/components/simulations/table-view/ScenariosTableView.tsx

```diff
@@ -0,0 +1,340 @@
+        case "timestamp":
+          return { ...col, cell: TimestampCell };
+        case "actions":
+          return { ...col, cell: ActionsCell };
+        default:
+          return col;
+      }
+    });
+  }, [projectSlug]);
+
+  // Create the store once using useRef - this is the proper Zustand pattern
+  // The store is created once and persists for the lifetime of the component
+  // urlSync: true enables automatic URL param sync (handled by store internally)
+  const storeRef = useRef<ReturnType<
+    typeof createDataGridStore<ScenarioRunRow>
+  > | null>(null);
+
+  if (storeRef.current === null) {
+    storeRef.current = createDataGridStore<ScenarioRunRow>({
+      columns: baseColumns,
+      defaultSorting: { columnId: "timestamp", order: "desc" },
+      getRowId: (row) => row.scenarioRunId,
+      storageKey: project?.id ? `scenarios-table-${project.id}` : undefined,
+      urlSync: true, // Enable automatic URL sync
+    });
+  }
+
+  // Use the store - this is the Zustand hook pattern
+  const useStore = storeRef.current;
+
+  // Subscribe to specific state slices to avoid re-renders on every state change
+  const filters = useStore((state) => state.filters);
+  const sorting = useStore((state) => state.sorting);
+  const page = useStore((state) => state.page);
+  const pageSize = useStore((state) => state.pageSize);
+  const globalSearch = useStore((state) => state.globalSearch);
+  const visibleColumns = useStore((state) => state.visibleColumns);
+  const groupBy = useStore((state) => state.groupBy);
+
+  // Get the full state for passing to DataGrid (but don't use it in effects)
+  const storeState = useStore();
+
+  // Fetch filtered scenario runs (ungrouped)
+  const {
+    data: scenarioData,
+    isLoading: isLoadingUngrouped,
+    isFetching: isFetchingUngrouped,
+    error: errorUngrouped,
+  } = api.scenarios.getFilteredScenarioRuns.useQuery(
+    {
+      projectId: project?.id ?? "",
+      filters: filters.map((f) => ({
+        columnId: f.columnId,
+        operator: f.operator,
+        value: f.value,
+      })),
+      sorting: sorting ?? undefined,
+      pagination: {
+        page,
+        pageSize,
+      },
+      search: globalSearch || undefined,
+      includeTraces: true,
+    },
+    {
+      enabled: !!project?.id && !groupBy,
+      refetchInterval: 30000,
+    }
+  );
+
+  // Fetch grouped scenario runs
+  const {
+    data: groupedData,
+    isLoading: isLoadingGrouped,
+    isFetching: isFetchingGrouped,
+    error: errorGrouped,
+  } = api.scenarios.getGroupedScenarioRuns.useQuery(
+    {
+      projectId: project?.id ?? "",
+      groupBy: groupBy ?? "",
+      filters: filters.map((f) => ({
+        columnId: f.columnId,
+        operator: f.operator,
+        value: f.value,
+      })),
+      sorting: sorting ?? undefined,
+      pagination: {
+        page,
+        pageSize,
+      },
+    },
+    {
+      enabled: !!project?.id && !!groupBy,
+      refetchInterval: 30000,
+    }
+  );
```

Comment:
This is questionable -- why is this done here instead of just intelligently by the controller or service layer?